server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -188,13 +178,27 @@ public void sendExecuteQuery(
             && (request.source() == null || request.source().rankBuilder() == null);
         Writeable.Reader<SearchPhaseResult> reader = fetchDocuments ? QueryFetchSearchResult::new : in -> new QuerySearchResult(in, true);
 
-        final ActionListener<? super SearchPhaseResult> handler = responseWrapper.apply(connection, listener);
+        final long startNanos = System.nanoTime();
         transportService.sendChildRequest(
             connection,
             QUERY_ACTION_NAME,
             request,
             task,
-            new ConnectionCountingHandler<>(handler, reader, connection)
+            new ConnectionCountingHandler<>(listener, reader, connection) {
```

Comment:
Removing at least one layer of listener indirection.
Also, inlining this logic here makes it obvious that the node id is never `null` (we use it as a key for the other map also so that would throw there).